File: frontend/src/services/inheritor.ts
import api from './api'

// === 类型定义 ===

export interface CategoryInfo {
  id: string
  name: string
  icon: string
  description: string
}

export interface ToolCapabilityPreview {
  tool_id: string
  tool_name: string
  description: string
  sample_output: string
}

export interface GeneratePersonaRequest {
  name: string
  category: string
  personality: string
  bio: string
  selected_tools: string[]
  expertise: string[]
}

export interface GeneratePersonaResponse {
  persona: string
  greeting: string
  quick_questions: string[]
  style: string
  suggested_avatar_prompt: string
}

export interface CustomInheritorCreate {
  name: string
  category: string
  persona: string
  greeting: string
  tools: string[]
  domain_prompts: Record<string, string>
  style: string
  expertise: string[]
}

export interface CustomInheritorUpdate {
  name?: string
  persona?: string
  greeting?: string
  tools?: string[]
  domain_prompts?: Record<string, string>
  style?: string
  expertise?: string[]
  is_public?: boolean
}

export interface CustomInheritor {
  id: number
  user_id: number
  name: string
  category: string
  avatar_url: string
  persona: string
  greeting: string
  tools: string[]
  domain_prompts: Record<string, string>
  style: string
  expertise: string[]
  is_public: boolean
  created_at: string
}

export interface InheritorStats {
  total: number
  limit: number
  remaining: number
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  pages: number
}

export interface InheritorContext {
  id: string
  name: string
  avatar: string
  system_prompt: string
  style: string
  expertise: string[]
  greeting: string
  tools: string[]
  domain_prompts: Record<string, string>
  quick_questions: string[]
  is_custom: boolean
}

// === API 函数 ===

export async function getCatalog(): Promise<CategoryInfo[]> {
  const res = await api.get('/api/inheritors/catalog')
  return res.data
}

export async function generatePersona(req: GeneratePersonaRequest): Promise<GeneratePersonaResponse> {
  const res = await api.post('/api/inheritors/generate-persona', req, { timeout: 60000 })
  return res.data
}

export async function createInheritor(data: CustomInheritorCreate): Promise<CustomInheritor> {
  const res = await api.post('/api/inheritors', data)
  return res.data
}

export async function listMyInheritors(): Promise<CustomInheritor[]> {
  const res = await api.get('/api/inheritors')
  return res.data
}

export async function getInheritorStats(): Promise<InheritorStats> {
  const res = await api.get('/api/inheritors/stats')
  return res.data
}

export async function listPublicInheritors(page: number = 1, pageSize: number = 12): Promise<PaginatedResponse<CustomInheritor>> {
  const res = await api.get('/api/inheritors/public', { params: { page, page_size: pageSize } })
  return res.data
}

export async function getInheritorDetail(id: number): Promise<CustomInheritor> {
  const res = await api.get(`/api/inheritors/${id}`)
  return res.data
}

export async function updateInheritor(id: number, data: CustomInheritorUpdate): Promise<CustomInheritor> {
  const res = await api.put(`/api/inheritors/${id}`, data)
  return res.data
}

export async function deleteInheritor(id: number): Promise<{ message: string }> {
  const res = await api.delete(`/api/inheritors/${id}`)
  return res.data
}

export async function generateAvatar(id: number): Promise<{ avatar_url: string }> {
  const res = await api.post(`/api/inheritors/${id}/avatar`, {}, { timeout: 120000 })
  return res.data
}

export async function getInheritorContext(personaId: string): Promise<InheritorContext> {
  const res = await api.get(`/api/chat/inheritor/${personaId}`)
  return res.data
}

File: backend/app/schemas/inheritor.py
"""自定义传承人 schemas"""

from datetime import datetime
from pydantic import BaseModel, Field


# === 品类目录 ===

class CategoryInfo(BaseModel):
    id: str
    name: str
    icon: str = "🏮"
    description: str = ""


# === 工具预览 ===

class ToolCapabilityPreview(BaseModel):
    tool_id: str
    tool_name: str
    description: str
    sample_output: str = ""


# === 人设生成 ===

class GeneratePersonaRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=50, description="传承人名称")
    category: str = Field(..., min_length=1, max_length=50, description="非遗品类")
    personality: str = Field(default="", max_length=500, description="性格描述")
    bio: str = Field(default="", max_length=200, description="一句话简介")
    selected_tools: list[str] = Field(default_factory=list, description="选择的工具列表")
    expertise: list[str] = Field(default_factory=list, description="专长标签")


class GeneratePersonaResponse(BaseModel):
    persona: str  # system_prompt
    greeting: str
    quick_questions: list[str]
    style: str
    suggested_avatar_prompt: str


# === 自定义传承人 ===

class CustomInheritorCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=50)
    category: str = Field(..., min_length=1, max_length=50)
    persona: str = Field(..., min_length=1)
    greeting: str = Field(..., min_length=1)
    tools: list[str] = Field(default_factory=list)
    domain_prompts: dict[str, str] = Field(default_factory=dict)
    style: str = Field(default="", max_length=200)
    expertise: list[str] = Field(default_factory=list)
    avatar_url: str = Field(default="", max_length=500)


class CustomInheritorUpdate(BaseModel):
    name: str | None = Field(None, max_length=50)
    persona: str | None = None
    greeting: str | None = None
    tools: list[str] | None = None
    domain_prompts: dict[str, str] | None = None
    style: str | None = Field(None, max_length=200)
    expertise: list[str] | None = None
    is_public: bool | None = None
    avatar_url: str | None = Field(None, max_length=500)


class CustomInheritorResponse(BaseModel):
    id: int
    user_id: int
    name: str
    category: str
    avatar_url: str = ""
    persona: str  # truncated to 100 chars in list context
    greeting: str
    tools: list[str] = []
    domain_prompts: dict[str, str] = {}
    style: str = ""
    expertise: list[str] = []
    is_public: bool = False
    created_at: datetime

    model_config = {"from_attributes": True}


class InheritorStats(BaseModel):
    total: int
    limit: int = 10
    remaining: int
